modify an occurrence

SharglutDev opened the pull request:
close #11468

SharglutDev committed:
front: replace editingTrainIsPacedTrain by editingItemType in store

Change the type from a boolean to an union type. This will facilitate handling occurrence edition.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: hide paced train switch when editing an occurrence

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: replace setItemIdToEdit by setItemToEditData

- pass occurrence id when editing an occurrence this will help us to track which paced train exception should be modified or added
- pass the current paced train to be able to compare which fields have been changed

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: upsert exceptions in paced train to update

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: update formatMargin to always return something

The back-end accept to receive undefined margins but if it does, will return the value `{ boundaries: [], values: ['0%'] }`.
This can lead to mismatch when trying to compare two timetable items, one that has already been passed by the back end and one not.
Now if margins have only one value of 0%, return it instead of undefined.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: fill occurrence with its exception and refacto useOccurrences

If an occurrence is an exception, update it with the exception fields so the user can see the difference in the timetable UI.
As each occurrence displays its rolling stock image, we need to handle the rolling stock fetching directly in useOccurrences.
To handle this async call to get the rolling stock details, refacto useOccurrences from a useMemo to a useState/useEffect design.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: fix exception category silo translation

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: display occurrence name in italic if start time exception

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

theocrsb commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -250,7 +255,7 @@ const PacedTrainItem = ({
           selectPathProjection={selectPathProjection}
           duplicateTimetableItem={duplicatePacedTrain}
-          editTimetableItem={() => selectPacedTrainToEdit(pacedTrain)}
+          editTimetableItem={() => selectPacedTrainToEdit(pacedTrain, pacedTrain)}
```
I find it a little confusing to pass 2 times `pacedTrain`

SharglutDev replied:
I can't really avoid to pass the second one because it's used to pass down the exceptions.
I added a comment.

theocrsb commented on front/src/reducers/osrdconf/operationalStudiesConf/__tests__/operationalStudiesConfReducer.spec.ts:
```diff
@@ -163,7 +165,7 @@ describe('simulationConfReducer', () => {
         timeWindow: new Duration({ minutes: 60 }),
         interval: new Duration({ minutes: 30 }),
-        editingTrainIsPacedTrain: true,
+        editingItemType: 'pacedTrain',
```
I think it would be a good idea to type this variable in order to have the type information on hovering

SharglutDev replied:
Done in a separate commit

theocrsb commented:
In some cases, translations are not displayed correctly. We've noticed that this problem seems a bit random.
![image](https://github.com/user-attachments/assets/d7b6e141-dbf6-48aa-9471-3a9a63d2de00)

anisometropie commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/__tests__/buildPacedTrainException.spec.ts:
```diff
@@ -0,0 +1,173 @@
+      updatedPacedTrainExceptions[0],
+      { ...updatedPacedTrainExceptions[1], key: '123-456' },
+    ]).toEqual(expectedPacedTrainExceptions);
+  });
```
why don’t you compare directly with what you want the function to return ?

 you can expect separately if that helps

expect(updatedPacedTrainExceptions[0])
expect(updatedPacedTrainExceptions[1])

SharglutDev replied:
Here, I know that the first exception will remain unchanged while the second one will have new changeGroups. Since its a new exception, the key will be generated so if I don't force pass one here the test won't pass the comparison.

Did I understand your comment or was it something else ?

SharglutDev replied:
Fixed

anisometropie commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/__tests__/buildPacedTrainException.spec.ts:
```diff
@@ -0,0 +1,173 @@
+    const occurrenceId = 'indexedoccurrence_123_2' as IndexedOccurrenceId;
+    const updatedPacedTrainExceptions = buildPacedTrainException(
+      originalPacedTrain,
+      pacedTrainWithOccurrenceChanges,
```
can’t you make the function’s api simpler by just passing what’s changed here in the second argument, not the whole train again ?

SharglutDev committed:
front: type the expected state for selectTrainToEdit() test

This helps to avoid any type mistake when build this state.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev commented:
> In some cases, translations are not displayed correctly. We've noticed that this problem seems a bit random.

Can't reproduce this issue anymore, did you restart your front container when switching on the branch ? (happened also to @anisometropie)

anisometropie commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -0,0 +1,163 @@
+ */
+export default function buildPacedTrainException(
+  originalPacedTrain: PacedTrain,
+  pacedTrainWithOccurrenceChanges: PacedTrain,
```
you could really avoid a lot of these comparisons if you find a way to pass just what changes, not the full paced train

SharglutDev replied:
Not really sure how I could do that. This would mean each time a field changes, I stock it somewhere (state, store) but only if we are updating an occurrence ? This seems painful.

anisometropie commented on front/src/reducers/osrdconf/types.ts:
```diff
@@ -52,7 +53,7 @@ export type OperationalStudiesConfState = OsrdConfState & {
   timeWindow: Duration;
   interval: Duration;
-  editingTrainIsPacedTrain: boolean;
+  editingItemType: 'trainSchedule' | 'pacedTrain' | 'occurrence';
```
better use an enum

emersion replied:
Could you expand why?

We've been using string unions in a lot of places already, and I'm not sure I understand why one should be preferred over the other. Both should provide the same level of type safety AFAIU.

anisometropie replied:
I would say it’s about the meaning they carry. 
Enums are a more idiomatic way to do that. having a string unions is as immediately obvious as for our intent.

When you read some code like
`buildException(train, train, 'trainSchedule')` you would have no idea it’s belonging to a limited set of values.

if you read `buildException(train, train, EditedItemType.trainSchedule)` 
the code is more expressive. the more named thing (function, variables) we can have the more the code is telling what’s it’s doing

Moreover a string union only exists at compile time. Enum create things ready to be used at runtime.

clarani replied:
I think the 2 propositions are valid, I just find the string enum more readable.

anisometropie commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/__tests__/buildPacedTrainException.spec.ts:
```diff
@@ -0,0 +1,191 @@
+    });
+  });
+
+  it('should update an existing exception if it was already present', () => {
```
I think both `it` l60 and l96 belong under the describe `describe('when the user modifies occurrence at index 2'`

SharglutDev replied:
Actually for this test (l.96), the user is editing an added occurrence so it's not the one at index 2

anisometropie replied:
you’re right, can you add another describe to give that info?

SharglutDev replied:
Done

theocrsb commented:
If you modify the category of an occurrence, and then modify the model train to match it, you won't reset the exception.

theocrsb reviewed (changes requested):
If a modification is made in NGE, exceptions are reset :

https://github.com/user-attachments/assets/800d0e1a-5c6c-40cf-8a7b-de34dfe36785

anisometropie committed:
front: handle exception reset when updating the paced train

Signed-off-by: SharglutDev <p.filimon75@gmail.com>
Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

anisometropie commented:
> If you modify the category of an occurrence, and then modify the model train to match it, you won't reset the exception.

Just made a commit to fix that with @theocrsb
https://github.com/OpenRailAssociation/osrd/pull/11887/commits/5bee1927909eb3ed650781549fdd19b95d025db5

SharglutDev committed:
front: keep paced train exceptions when updating one in nge

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev commented:
> If a modification is made in NGE, exceptions are reset 

Fixed in a separate commit

SharglutDev commented:
> > If you modify the category of an occurrence, and then modify the model train to match it, you won't reset the exception.
> 
> Just made a commit to fix that with @theocrsb [5bee192](https://github.com/OpenRailAssociation/osrd/commit/5bee1927909eb3ed650781549fdd19b95d025db5)

oops I didn't noticed you added a commit, I pushed one wip to fix that, I'll compare tomorrow what we both have done

SharglutDev committed:
front: fix null category not handled in the occurrence indicator

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: handle exception reset when updating in nge

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

anisometropie commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -403,9 +404,30 @@ const handleUpdateTimetableItem = async ({
         : DEFAULT_TIME_WINDOW.toISOString(),
     };
+
+    if (!('exceptions' in timetableItem)) {
```
`in` only checks if the key exists, it would also be true if the value is `undefined` or `null`. (which exceptions is not supposed to be)

SharglutDev replied:
This was the only solution I have to be able to do `timetableItem.exception` as checking in the previous if the `trainrun.frequencyId` doesn't narrow down the type of `timetableItem` to a `PacedTrain`. Do you think of something else we could do ?

emersion replied:
`isPacedTrain()`?

SharglutDev replied:
Oh you're right, but it's `isPacedTrainResponseWithPacedTrainId` function, forgot we have this one

SharglutDev replied:
Pushed the fix

anisometropie commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -38,83 +36,130 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
+          exceptions,
+          occurrenceId
+        );
+        const correspondingException = removeNullProps(correspondingExceptionWithNullChangeGroups);
```
I guess we can remove it if backend no longer sends us `null` values

SharglutDev replied:
The first back PR dealing with that didn't work but it has been fixed now. Removed the commit :)

anisometropie commented on front/src/modules/trainschedule/components/Timetable/Timetable.tsx:
```diff
@@ -35,9 +40,9 @@ type TimetableProps = {
+  setItemToEditData: (itemToEditData?: ItemToEditData) => void;
   removeTimetableItems: (timetableItemsToRemove: TimetableItemId[]) => void;
-  itemIdToEdit?: TimetableItemId;
+  itemToEditData?: ItemToEditData;
```
“item” lacks expliciteness, doesn’t tell us what it is we’re editing.

`editedTrainData` or something like that ?

SharglutDev replied:
Well we decided to use item for train/paced train in the timetable. Maybe I could use `timetableItemEditData` ?

SharglutDev replied:
Done

anisometropie commented:
when editing one occurrence start time, and then back to it’s original occurrence start time, it doesn’t remove the exception status. still shows as a start time exception

SharglutDev committed:
front: fix destination stop duration in nge

In micro, to make the train stop at its destination, we add a stopFor 0 on the last pathStep.
In nge side, as the destination doesn't have any arrival or departure by default, we were erasing this schedule.
This was an issue if the paced train had been modified one time in nge, thus erasing the default schedule behavior, because to compute a paced train exception we compare the schedule of the occurrence being modified (which gets the stopFor 0 on last pathStep) and the one from the paced train.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

theocrsb commented:
> when editing one occurrence start time, and then back to it’s original occurrence start time, it doesn’t remove the exception status. still shows as a start time exception

I don't have this error, can you try again?

SharglutDev commented:
> when editing one occurrence start time, and then back to it’s original occurrence start time, it doesn’t remove the exception status. still shows as a start time exception

Can't reproduce. Or did I misunderstood the issue ?


https://github.com/user-attachments/assets/5574b540-54b9-47f2-8083-4bfc218f7b8c

SharglutDev commented:
> when editing one occurrence start time, and then back to it’s original occurrence start time, it doesn’t remove the exception status. still shows as a start time exception

Ok reproduced, I pushed a fix

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -38,83 +35,122 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
+          trainName: correspondingException?.train_name?.value ?? computeOccurrenceName(name, i),
+          rollingStock: occurrenceRollingStock,
+          startTime: correspondingException?.start_time?.value
+            ? new Date(correspondingException?.start_time?.value)
```
```suggestion
            ? new Date(correspondingException.start_time.value)
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -38,83 +35,122 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
+          stopsCount: correspondingException?.path_and_schedule
+            ? (correspondingException?.path_and_schedule.schedule?.filter(
+                (step) => step.stop_for && Duration.parse(step.stop_for).ms > 0
+              ).length ?? 0) + 1 // +1 to take the final stop (destination) into account,
```
- thank you for the comment, it's clearer with it
- even if the duration of the stop is 0 ms, the train will stop so I think we should consider it as a stop. Maybe we can remove the second part of the test ?
```suggestion
            ? (correspondingException.path_and_schedule.schedule.filter((step) => step.stop_for)
                .length ?? 0) + 1 // +1 to take the final stop (destination) into account,
```

SharglutDev replied:
Hmm if we remove this condition, the destination will be counted twice most of the time as we instantiate it with a `stopFor: 0` so we need to remove the `+ 1` at the end. I think this 0 stop is a hack we use to make the train stop at the end of its itinerary but making a stop: 0 in the middle of the itinerary doesn't make sense to me irl ?

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -38,83 +35,122 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
-      });
-    }
-
-    // Handle added exceptions
```
I liked this comment :)

SharglutDev replied:
Of my bad, it shouldn't have been removed ! Probably a rebase issue

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -38,83 +35,122 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
+
+        computedOccurrences.push({
+          id: formatPacedTrainIdToExceptionId(id, exception.key),
+          trainName: exception.train_name?.value ?? `${name}/+`,
```
Why do you have the `/+` ? It's an AC ?

SharglutDev replied:
Oh yes, it has been proposed on matrix by @maelysLeratRosso when we were thinking how to generate added exceptions name. It's more a feature from this [PR](https://github.com/OpenRailAssociation/osrd/pull/11848), but I added it here as well to facilitate @anisometropie rebase as we decided to merge mine before his.

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/OccurrenceItem.tsx:
```diff
@@ -98,9 +94,7 @@ const OccurrenceItem = ({
-        occurrenceRollingStock = await getRollingStockByName({
-          rollingStockName: occurrenceToUpdateException.rolling_stock.rolling_stock_name,
-        }).unwrap();
+        occurrenceRollingStock = occurrence.rollingStock;
```
I think this should be enough now :thinking: 
```
updatedPacedtrain = {
        ...pacedTrainWithOccurrenceDetails,
        rollingStock: occurrence.rollingStock,
};
```

SharglutDev replied:
Not sure what you mean ? If the occurrence doesn't have a rs exception we want to keep the one from the paced train 🤔

clarani replied:
Ok we can keep this for later :+1:

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/OccurrenceItem.tsx:
Can you add a TODO to refacto the function editOccurrence ? Actually, we can simplify the code and get more information from `occurrence`.

`editOccurrence` can be only this: 
```
  const editOccurrence = async () => {
    const updatedPacedTrain = formatPacedTrainWithOccurenceDetails(currentPacedTrain, occurrence);
    selectPacedTrainToEdit(updatedPacedTrain, currentPacedTrain, occurrence.id);
  };
```
(but this can be done later)

SharglutDev replied:
But not sure to understand your thoughts here, we should discuss about that :)
You mean you would prefer to handle everything inside `formatPacedTrainWithOccurenceDetails` ?

SharglutDev replied:
Done and created an [issue](https://github.com/OpenRailAssociation/osrd/issues/12030)

clarani replied:
Thanks :pray:

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -80,7 +80,9 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
           disabled: correspondingException?.disabled,
-          category: correspondingException?.rolling_stock_category?.value ?? pacedTrainCategory,
+          category: correspondingException?.rolling_stock_category
+            ? correspondingException?.rolling_stock_category.value
```
```suggestion
            ? correspondingException.rolling_stock_category.value
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/OccurrenceIndicator.tsx:
```diff
@@ -97,7 +97,7 @@ const OccurrenceIndicator = ({ occurrence }: OccurrenceIndicatorProps) => {
                       ? t(`occurrenceChangeGroup.${changeGroup}`)
                       : t(
-                          `rollingStock.categoriesOptions.${occurrence.exceptionChangeGroups?.rolling_stock_category?.value}`,
+                          `rollingStock.categoriesOptions.${occurrence.exceptionChangeGroups?.rolling_stock_category?.value ?? 'noCategory'}`,
```
Nice :)

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -80,7 +80,9 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
             : stopsCount,
           disabled: correspondingException?.disabled,
-          category: correspondingException?.rolling_stock_category?.value ?? pacedTrainCategory,
+          category: correspondingException?.rolling_stock_category
+            ? correspondingException?.rolling_stock_category.value
```
Hum I'm afraid we break this in the future. Do you think you can add a small comment here ?

SharglutDev replied:
Sure, done.

clarani commented on front/src/reducers/osrdconf/operationalStudiesConf/trainSettingsReducer.ts:
```diff
@@ -61,8 +61,11 @@ const trainSettingsReducer = {
   },
-  toggleEditingTrainIsPacedTrain(state: Draft<OperationalStudiesConfState>) {
-    state.editingTrainIsPacedTrain = !state.editingTrainIsPacedTrain;
+  updateEditingItemType(
```
What about creating a new action `toggleEditingItemType` instead of `updateEditingItemType` ?
```
toggleEditingItemType(state: Draft<OperationalStudiesConfState>) {
  if (state.editingType === 'occurrence') throw new Error();
  state.editingItemType = state.editingItemType === 'pacedTrain' ? 'trainSchedule' : 'pacedTrain'
},
```

SharglutDev replied:
Done

anisometropie replied:
wait , we can’t have an occurrence here being edited ? what happens when user edits an occurrence ?

SharglutDev replied:
The editing type is updated in the store by the `selectTrainToEdit` directly. That's the only spot where it becomes `occurrence`

clarani commented on front/src/modules/trainschedule/helpers/updateTimetableItemHelpers.ts:
```diff
@@ -141,29 +142,29 @@ export async function storeTrainSchedule(
 
 export async function storePacedTrain(
-  timetableItemIdToUpdate: TimetableItemId,
+  timetableItemIdToUpdate: TimetableItemToEditData,
```
Why do you pass the TimetableItemToEditData here ? You only use the timetableItemId
```suggestion
  timetableItemIdToUpdate: TimetableItemId,
```

SharglutDev replied:
You're right, done

clarani commented on front/src/reducers/osrdconf/types.ts:
```diff
@@ -161,6 +162,11 @@ export type PacedTrainId = string & { readonly __type: unique symbol };
 
 export type TrainId = TrainScheduleId | OccurrenceId;
 export type TimetableItemId = TrainScheduleId | PacedTrainId;
+export type TimetableItemToEditData = {
+  timetableItemId: TimetableItemId;
+  originalPacedTrain?: PacedTrainWithDetails;
+  occurrenceId?: OccurrenceId;
+};
```
Hum I've thinking, it might be better to have this type, what do you think ?
```suggestion
export type TimetableItemToEditData =
  | {
      timetableItemId: TrainScheduleId;
    }
  | {
      timetableItemId: PacedTrainId;
      originalPacedTrain: PacedTrainWithDetails;
      occurrenceId?: OccurrenceId;
    };
```

SharglutDev replied:
Done. Had to update some stuff (in the `upsert exceptions in paced train to update` commit) in `formatPacedTrainPayload` to narrow down the proper kind of `timetableItemToEditData`

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -251,7 +255,9 @@ const PacedTrainItem = ({
-          editTimetableItem={() => selectPacedTrainToEdit(pacedTrain)}
+          // The first pacedTrain is the one we use to update it
+          // The second one is the one we use to pass down the exceptions
+          editTimetableItem={() => selectPacedTrainToEdit(pacedTrain, pacedTrain)}
```
When you edit the pacedTrain, I think you don't need to give the original one
```suggestion
          editTimetableItem={() => selectPacedTrainToEdit(pacedTrain)}
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/Timetable.tsx:
```diff
@@ -124,9 +129,17 @@ const Timetable = ({
+        timetableItemId: itemToEdit.id,
+        originalPacedTrain,
+        occurrenceId,
+      });
```
With the new type:
```
const selectTimetableItemToEdit = useCallback(
  (itemToEdit: TimetableItemWithDetails, occurrenceId?: OccurrenceId) => {
    dispatch(selectTrainToEdit({ item: itemToEdit, isOccurrence: !!occurrenceId }));
    setTimetableItemToEditData(
      isPacedTrainWithDetails(itemToEdit)
        ? { timetableItemId: itemToEdit.id, originalPacedTrain: itemToEdit, occurrenceId }
        : { timetableItemId: itemToEdit.id }
    );
    setDisplayTrainScheduleManagement(MANAGE_TRAIN_SCHEDULE_TYPES.edit);
  },
  []
);
```
And you can avoid to give the paced train twice

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/hooks/useUpdateTimetableItem.ts:
```diff
@@ -51,7 +56,7 @@ const useUpdateTimetableItem = (
   return async function submitConfUpdateTrainSchedules() {
     if (
-      !timetableItemIdToEdit ||
+      !timetableItemToEditData ||
```
You can extract timetableItemId just after this if

SharglutDev replied:
Sure, done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/TimetableManageTrainSchedule.tsx:
```diff
@@ -75,41 +79,42 @@ const TimetableManageTrainSchedule = ({
-                      updateEditingItemType(
-                        editingItemType === 'pacedTrain' ? 'trainSchedule' : 'pacedTrain'
+        {displayTrainScheduleManagement === MANAGE_TRAIN_SCHEDULE_TYPES.edit &&
+          timetableItemToEditData?.timetableItemId && (
```
```suggestion
          timetableItemToEditData && (
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -34,18 +40,66 @@ export function formatTimetableItemPayload(
+    },
+    path: pacedTrainWithDetails.path,
+    power_restrictions: pacedTrainWithDetails.power_restrictions,
+    rolling_stock_name: pacedTrainWithDetails.rollingStock!.name,
```
Why rollingStock is not undefined here ? If it comes from NGE ? or is it '' ?

SharglutDev replied:
Well spotted, at that point I didn't handle nge. That's why on another commit, I added the condition `timetableItemToEditData.originalPacedTrain.rollingStock` to make sure it's defined before calling the function and a comment to explain why

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -0,0 +1,175 @@
+ */
+export default function updatePacedTrainException(
+  originalPacedTrain: PacedTrain,
+  pacedTrainWithOccurrenceChanges: PacedTrain,
```
Maybe we can rename this field occurrence ?

SharglutDev replied:
I chose this naming because we still handle a `PacedTrain` type here, I didn't want to lose this from perspective. Would it be ok for you to leave it like that ?

SharglutDev replied:
Done, used "updatedOccurrence"

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -34,18 +40,66 @@ export function formatTimetableItemPayload(
   const baseTrain = formatTimetableItemPayload(osrdconf, rollingStockName);
-  return {
+
+  let pacedTrainToUpdate: PacedTrain = {
```
What about newPacedTrain ?
```suggestion
  let newPacedTrain: PacedTrain = {
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -34,18 +40,66 @@ export function formatTimetableItemPayload(
+    const originalPacedTrain = formatPacedTrainWithDetailsToPacedTrainPayload(
+      timetableItemToEditData.originalPacedTrain
+    );
+    const updatedExceptions = updatePacedTrainException(
```
This function is quite heavy and mix 2 different steps. Do you think it's possible to split it into 2 functions ? One that will generate the new exception and one that will update the exceptions array ?
Something like this should work:
```
const exception = generateException(originalPacedTrain, newPacedTrain, occurrenceId);

const updatedExceptions = updatePacedTrainExceptions(
  originalPacedTrain.exceptions,
  exception,
  occurrenceId
);

return {
  ...originalPacedTrain,
  exceptions: updatedExceptions,
};
```

SharglutDev replied:
Sure, done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -30,6 +30,8 @@ export default function updatePacedTrainException(
     occurrenceId
   );
 
+  // TROUVER COMMENT GERER SANS OCCURRENCE ID > boucler sur les exceptions d'origine et voir s'il reste des exceptions
```
To remove ;)

SharglutDev replied:
Oops ! Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -173,3 +175,126 @@ export default function updatePacedTrainException(
+ * If the paced train matches an exception change group value and if it was the last one remaining,
+ * the exception is removed.
+ */
+export function checkChangeGroups(
```
Maybe this function could have its own file :thinking:

SharglutDev replied:
As all these functions were about exceptions and have some similarities, I left them in the same file. But if you really think the file is too big I can move them :)

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -173,3 +175,126 @@ export default function updatePacedTrainException(
+ * the exception is removed.
+ */
+export function checkChangeGroups(
+  originalExceptions: PacedTrainException[],
+  updatedPacedTrain: PacedTrain
```
very nit: I find it more natural to give the updatedPacedTrain before the originalExceptions (but it's very nit)
```suggestion
  updatedPacedTrain: PacedTrain,
  originalExceptions: PacedTrainException[]
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -173,3 +175,126 @@ export default function updatePacedTrainException(
+ */
+export function checkChangeGroups(
+  originalExceptions: PacedTrainException[],
+  updatedPacedTrain: PacedTrain
```
I think you can change the type of the PacedTrain, so we clearly understand that we are building the new exceptions array
```suggestion
  updatedPacedTrain: Omit<PacedTrain, 'exceptions'>,
```

SharglutDev replied:
`updatedPacedTrain` is only here for comparison. In the first line of the function, we can clearly see that we build a new exception array by looping on the original ones.

I don't really like changing the type here because this implies either : 
1. Removing the exceptions from the `updatedPacedTrain` before passing it to this function so more boilerplate
2. Leaving `updatedPacedTrain` as it is (since typescript doesn't complain) but we pass a `PacedTrain` and use inside an `Omit<PacedTrain, 'exceptions'>` which I think is weird

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -173,3 +175,126 @@ export default function updatePacedTrainException(
+    // We do the check only for indexed occurrences because added exceptions should not have
+    // their start time reset
+    if (exception.start_time && exception.occurrence_index !== undefined) {
+      let originalStartTimeToTest = new Date(updatedPacedTrain.start_time);
+
+      const originalPacedTrainInterval = Duration.parse(updatedPacedTrain.paced.interval);
+      originalStartTimeToTest = dayjs(originalStartTimeToTest)
+        .add(exception.occurrence_index * originalPacedTrainInterval.ms, 'ms')
+        .toDate();
```
You can simplify a bit like this:
```suggestion
      const originalStartTimeToTest = dayjs(new Date(updatedPacedTrain.start_time))
        .add(exception.occurrence_index * Duration.parse(updatedPacedTrain.paced.interval).ms, 'ms')
        .toDate();
```

SharglutDev replied:
It is indeed shorter but don't you think it is a bit painful to read ? :)

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -85,21 +91,40 @@ export function formatPacedTrainPayload(
 
-  if (timetableItemToEditData?.originalPacedTrain && timetableItemToEditData?.occurrenceId) {
+  // We need to check for rollingstock because when its missing it means we just created a train from
+  // nge or with import so the rs will be missing and we don't need to check for exceptions anyway (for now)
```
Hum I'm not sure about this. I think that even if the paced train has no rollingstock, I can still define some exceptions if I create a train from NGE and then edit one occurrence.

So we should check the exceptions even if the new paced train has no rollingStock

SharglutDev replied:
As occurrences in an invalid paced train have an "disabled" look, I thought we could not edit them (that's why I put the "for now" in the comment) and turns out we can edit them so you're right we should be able to generate exception in that case. I removed the if here and added a fallback to empty string for rolling stock name if we don't have a rs provided.

I also spotted a small bug in the `checkChangeGroups` function which I fixed

clarani replied:
Hum I think you didn't resolve this comment ?

SharglutDev replied:
Oh my bad, I actually resolved it (on the `upsert...` commit) but I ended but putting back the condition (in commit `handle exception reset...`) probably when resolving a rebase issue. Should be fixed.

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -85,21 +91,40 @@ export function formatPacedTrainPayload(
+        pacedTrainToUpdate
+      );
+
+      pacedTrainToUpdate = {
+        ...pacedTrainToUpdate,
+        exceptions: updatedExceptions,
+      };
```
Add early return
```
return {
      ...newPacedTrain,
      exceptions: checkChangeGroups(newPacedTrain, originalPacedTrain.exceptions),
    };
```

SharglutDev replied:
Not sure where you want this early return ?

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -403,12 +411,22 @@ const handleUpdateTimetableItem = async ({
         : DEFAULT_TIME_WINDOW.toISOString(),
     };
+
+    if (!isPacedTrainResponseWithPacedTrainId(timetableItem)) {
+      throw new Error('Paced train must have an exceptions property');
+    }
```
Hum what if I transform a trainSchedule into a PacedTrain ? timetableItem will be a TrainScheduleItem I think

SharglutDev replied:
Oh you're right, didn't remember we could do that in nge as well. Done

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -403,12 +411,22 @@ const handleUpdateTimetableItem = async ({
+      throw new Error('Paced train must have an exceptions property');
+    }
+
+    const updatedPacedTrain: PacedTrain = {
+      ...timetableItemForUpdate,
+      paced,
+      exceptions: timetableItem.exceptions,
+    };
```
It's weird to give an array of exceptions to erase it immediately 
```
    const basePacedTrain = { ...timetableItemForUpdate, paced };
    const updatedPacedTrain: PacedTrain = {
      ...basePacedTrain,
      exceptions: checkChangeGroups(basePacedTrain, timetableItem.exceptions),
    };
```

SharglutDev replied:
Agreed, if I do that, is it better for you ?
```
const updatedPacedTrain: PacedTrain = {
      ...timetableItemForUpdate,
      paced,
      exceptions: [],
    };

    updatedTimetableItem = await storePacedTrain(
      timetableItemId,
      {
        ...updatedPacedTrain,
        exceptions: isPacedTrainResponseWithPacedTrainId(timetableItem)
          ? checkChangeGroups(timetableItem.exceptions, updatedPacedTrain)
          : [],
        [...]
```

SharglutDev replied:
Should be fixed

theocrsb reviewed (approved):
LGTM and tested

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/buildPacedTrainException.ts:
```diff
@@ -0,0 +1,185 @@
+/**
+ * Compare the original paced train with the one from the occurrence update and
+ * fill the original paced train exceptions property every time a field is different.
+ * Add or update the exception in the exceptions list.
```
This description is not accurate anymore

SharglutDev replied:
Oh right, I did remove the last line but on a later commit. Fixed.

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -97,37 +101,47 @@ export function formatPacedTrainPayload(
-      originalPacedTrain,
-      timetableItemToEditData.occurrenceId
-    );
+    if (timetableItemToEditData?.occurrenceId) {
```
```suggestion
    if (timetableItemToEditData.occurrenceId) {
```

SharglutDev replied:
Done

clarani commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -54,7 +55,11 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
           .toDate();
 
-        const correspondingException = findExceptionWithOccurrenceId(exceptions, occurrenceId);
+        const correspondingExceptionWithNullChangeGroups = findExceptionWithOccurrenceId(
```
Do you want to remove this commit ?

SharglutDev replied:
I didn't remove it because the first back PR wasn't working. A [fix](https://github.com/OpenRailAssociation/osrd/pull/12068) is passed now, I'll retest it :)

SharglutDev replied:
Works great now ! Removed the commit :)

clarani commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -406,12 +414,20 @@ const handleUpdateTimetableItem = async ({
         : DEFAULT_TIME_WINDOW.toISOString(),
     };
+
+    const updatedPacedTrain: PacedTrain = {
+      ...timetableItemForUpdate,
+      paced,
+      exceptions: [],
+    };
```
nit:
```suggestion
const basePacedTrain = { ...timetableItemForUpdate, paced };
    const updatedPacedTrain = {
      ...basePacedTrain,
      exceptions: isPacedTrainResponseWithPacedTrainId(timetableItem)
        ? checkChangeGroups(basePacedTrain, timetableItem.exceptions)
        : [],
    };
```

And then:
```
 updatedTimetableItem = await storePacedTrain(
      timetableItemId,
      updatedPacedTrain,
      timetableId,
      dispatch,
      addUpsertedTimetableItems,
      addDeletedTimetableItemIds
    );
```

SharglutDev replied:
Done

clarani reviewed (approved):
LGTM :white_check_mark: nice work :muscle:
